### PR TITLE
Improve performance of length on UTF8String and UTF16String

### PR DESCRIPTION
**Pull request by ScottPJones**
This adds an optimized version to get the length of a UTF16String, it is about an order of magnitude faster (it is still O(n) like UTF8String, instead of O(1) like ASCIIString or UTF32String, though)

### CODE REVIEW

File: src/support/utf8.c [hunk 1]
```diff
@@ -232,7 +232,14 @@ size_t u8_offset(const char *s, size_t charnum)
...
 }
 
-/* byte offset => charnum */
+/**
+ * @brief      Calculates number of characters in a UTF-8 string
+ * 
+ * @param[in]  iStr UTF-8 encoded string
+ * @param[in]  iLen Length of the string in bytes
+ *
+ * @return number of logical characters
```
Comment (nalimilan) [inline src/support/utf8.c]: Probably better to say "codepoints" instead of "logical character".

File: src/support/utf8.c [hunk 2]
```diff
@@ -232,7 +232,14 @@ size_t u8_offset(const char *s, size_t charnum)
     return i;
 }
 
-/* byte offset => charnum */
+/**
+ * @brief      Calculates number of characters in a UTF-8 string
+ * 
+ * @param[in]  iStr UTF-8 encoded string
```
Comment (nalimilan) [inline src/support/utf8.c]: The parameter is called `s`, not `iStr`. The convention in this file seems to be to use short lower case identifiers anyway.

**Comment by nalimilan** [comment]
Do you really need to rely on a C function to do that? I'm not sure why this is currently the case for UTF-8, but I think it may have to do with the special status of `UTF8String`, which shouldn't be an issue for `UTF16String`.

**Comment by ScottPJones** [comment]
It is for speed, it simply was not nearly fast enough in Julia... about 3x slower... the Julia team will need to look at that, but for now, given that u8_charnum is already there, I don't think it should prevent this from being merged in.

File: src/support/utf8.c (hunk 1 above)
Comment (ScottPJones) [inline src/support/utf8.c]: At least I documented it! ;-)  From where I'm coming from, I thought logical character was actually clearer... a lot of people get codepoints vs. the value of the byte/word/32-bit confused.

File: src/support/utf8.c (hunk 2 above)
Comment (ScottPJones) [inline src/support/utf8.c]: Yes - I had moved u8_charnum (which I'd already rewritten and which has already been moved into 0.4 and 0.3.8) to my own utf.h / utf.c files, with my conversion speedups written in C also,
but then people wanted it in pure Julia, so I separated out that part, and moved the new u8_charnum and u16_charnum back into utf8.c... forgot to update the doxygen comment for that function, thanks for catching that!

Comment (ScottPJones) [inline src/support/utf8.c]: In my own utf.h/utf.c, I was following the convention agreed upon at the Belgian startup I'm consulting for... we are using Julia, at least in parts (but increasingly, people want to use it everywhere ;-) )

**Comment by ScottPJones** [comment]
@nalimilan Better now?  Thanks very much for the input!

**Comment by vtjnash** [comment]
> It is for speed, it simply was not nearly fast enough in Julia... about 3x slower...

[citation needed]

**Comment by carnaval** [comment]
We should definitely be able to get a simple loop like that down to the same IR as, e.g., clang would. If not I'd consider that a compiler bug.

**Comment by ScottPJones** [comment]
@carnaval I'm all for that, and I've said earlier, that the moment the Julia version is within say 5-10% of the C version, then the C version could be dropped... Unfortunately, that is *not* the case at present, and the people who would be able to improve it are probably working 90+ hour weeks as it is!
I need decent performance for these things, if I intend to use Julia for my string processing / database tasks... I don't think it is unreasonable to have some C code to get that performance, *at least for the time being*.

**Comment by ScottPJones** [comment]
@vtjnash I'll try to dig that code up, or recreate it... as I've said before, I am not opposed to having everything all written in Julia... but I need the performance now, not next year, and I don't think it is unreasonable to add a few lines of C to Julia, that are exactly like the code that is already there,
but for UTF-8 [I actually was about to send a message to @JeffBezanson about removing some C code from utf8.c, that is no longer used, and potentially unsafe..., so the net amount of C code in Julia would actually go down, not up!]

**Comment by carnaval** [comment]
(disclaimer : I know nothing about string encoding)
Using a literal translation of your C code I get equivalent speed (sometimes a bit faster, sometimes a bit slower) between the C version and this
```julia
@inline is_surrogate_trail(c :: UInt16) = (c & ~UInt16(0x3ff)) == UInt16(0xdc00)
function length2(s::UTF16String)
    d = s.data
    len = length(d) - 1
    len == 0 && return 0
    cnum = 0
    for i = 1:len
        @inbounds cnum += !is_surrogate_trail(d[i])
    end
    cnum
end
```
Note : my LLVM version (3.6) auto-vectorizes this nicely so it may be that your C compiler  is doing it but not the LLVM version you are using with julia.

**Comment by ScottPJones** [comment]
@carnaval This is what I got: for 64K strings, the Julia code was 2-2.2x slower, for 16 byte strings, it was between 41-72% slower... that's a very significant difference... The difference was < 3x now, because my Julia code is better than it was a couple of weeks ago ;-)

**Comment by ScottPJones** [comment]
@carnaval That's good... but I couldn't find the `@inline` macro in the documentation for v0.3.x... is that new to v0.4? 
Also, about the LLVM, I have a feeling that not everybody is up to that version... here is the version I have with OS X 10.10.4:
>LLVM (http://llvm.org/):
  LLVM version 3.3
  Optimized build.
  Built Apr 27 2015 (22:46:30).
  Default target: x86_64-apple-darwin14.3.0
  Host CPU: core-avx2

Here is a gist with my results and (really bad) benchmarking code...
https://gist.github.com/ScottPJones/bb712f7b85d1d8d91a9a

[I really want to know how in Julia to best to get the 2 numbers from `@time`, save them for a whole set of results, and then load them back in to calculate the differences in timings and memory utilization... I miss having a multidimensional associate array structure, both persistent and local, to do that with a single command..]

**Comment by carnaval** [comment]
As I said, you have to check if your C version is getting vectorized (likely) for example by running `gdb src/support/utf8.o -ex 'disassemble /m u16_charnum'`, and if it is the case for your julia version, for example by running `code_native(length, (UTF16String,))` or `code_llvm` if you prefer reading IR than assembly.

**Comment by ScottPJones** [comment]
@carnaval I would think it's the same LLVM... the C code is in utf8.c and is built as part of Julia, or have I missed something?

**Comment by carnaval** [comment]
And now that I look at your code, I'm pretty sure that the bound checks are not being ellided which prevents vectorization because of the possibility of diverging control flow. See the `@inbounds` in my example (which again, performs exactly as the C version on my machine).

**Comment by carnaval** [comment]
There is no particular reason for the C source to be compiled by clang using the same LLVM version we use for the codegen. Any C compiler does the job.

**Comment by ScottPJones** [comment]
@carnaval OK, very helpful!  As I think I've said, my very first attempt at this *was* in Julia, but then I saw that it was still slower than the UTF-8 version (and should have been faster), and enough slower that I felt I wouldn't be able to use it... and also, @JeffBezanson had it as a C function, which I had already sped up last week by a good amount, and if it was good enough for somebody like him, one of the Jedi Knights of Julia, I thought it *had* to be good enough for me!

**Comment by ScottPJones** [comment]
@carnaval Also good to know, that it won't necessarily be using the same LLVM as Julia... I often am running with the pre-release Xcode versions on my machine... Thanks!

**Comment by ScottPJones** [comment]
@carnaval Thanks so much... I just ran with your @inline and @inbound change, with my full set of different types of Unicode strings (i.e. only ASCII characters, some Latin1 or other 2-byte ones (for UTF8), 3-byte and 4 byte [i.e. surrogates for UTF-16).  It is 9-19% slower than the C version for 16 character strings... I need to test with more sizes I think... but it is 12-37% faster for 64K strings...

**Comment by ScottPJones** [comment]
My question now to the peanut gallery... since this is a mixed bag, should I change this to the pure Julia version, it is a bit slower than what I'd considered to be my pain threshold, but only for short strings... (but those are more common), and put my faith in the Julia team that they will look into that case... [it's a challenge for them], or see if the C version should be accepted for now?
If you all want to merge in the Julia version, shall I check it in with the C versions of both u8_charnum and u16_charnum removed?

Thanks so much for the help!

**Comment by vtjnash** [comment]
any idea why the `@inline` annotation is needed? or was just `@inbounds` needed? the inlining heuristics shouldn't have had any issues with such a simple function

yes, the julia-only version can be merged

**Comment by ScottPJones** [comment]
@carnaval About the Julia code, I'd just copied that from somebody else's Julia code in one of the utfXX.jl modules, and just changed the testing in the inner loop [to simply assume that it was valid UTF-8 or UTF-16, and then simply not count the trailing surrogates or continuation bytes]
Now I wonder if a lot of those loops I saw in the utfXX.jl modules should be liberally sprinkled with
`@inbounds`.

**Comment by vtjnash** [comment]
(note, i believe that `is_surrogate_trail` is already implemented as `utf16_is_trail`)

**Comment by ScottPJones** [comment]
@vtjnash I haven't had time to test all the permutations... but I'm very excited at the results!
This is more like what I was hoping to see out of Julia...

**Comment by ScottPJones** [comment]
@vtjnash I made a copy so I could put the `@inline` on it... I don't know if that works outside a module... and also I wanted to avoid the time taken to rebuild Julia... too excited to wait!

**Comment by vtjnash** [comment]
> Now I wonder if a lot of those loops I saw in the utfXX.jl modules should be liberally sprinkled with
@inbounds.

yes, if it helps performance. otherwise, no, since it can make bugs much harder to find. Automated bounds checking elision is something that is very difficult for a compiler, but not too difficult for a human.

**Comment by ScottPJones** [comment]
Also, I think I'll just replace the UTF16String version of length, because the old Julia code was always much worse than this... I'll leave the optimized UTF8String version for now, so as not to inadvertently slow things down, right after the change I put in last week to speed it up!

**Comment by ScottPJones** [comment]
@vtjnash :-) You don't know me yet... but I am a bit OCD about rigorously benchmarking stuff... of course I'll only put an @inbounds if it noticeably affects performance...

**Comment by JeffBezanson** [comment]
We should avoid `@inbounds` when dealing with possibly-invalid data. One more reason I support only allowing valid data in strings: it's more rigorous *and* performs better.

File: src/support/utf8.c (hunk 1 above)
Comment (nalimilan) [inline src/support/utf8.c]: "logical character" can mean either code point or grapheme cluster, I don't think it's clearer. A Google search for the former does not give anything useful, while a search for the latter immediately leads to the definition of Wikipedia.

**Comment by nalimilan** [comment]
@JeffBezanson What could possibly go wrong with corrupt data here? Since the corruption cannot affect the length of `s.data`, only its contents, could it really create problems with `@inbounds`?

**Comment by ScottPJones** [comment]
@JeffBezanson Since I am already checking the bounds, by using the length of s.data, what could possibly be the problem???

**Comment by ScottPJones** [comment]
@JeffBezanson Also, for the conversion code that I wrote (which I now have to update again and re-benchmark, after learning about `@inbounds` and `@inline`), it also knows exactly how long the array is.
I wonder - could the compiler be smart enough to see that I am using the bounds of the array for the loop, and simply act as if I used `@inbounds` automatically?  Now that would be a really useful optimization!

**Comment by ScottPJones** [comment]
Thanks everybody for all the help, I have now changed it, following @carnaval's great advice, to be pure Julia, not only for UTF16String, but also for UTF8String.  The people responsible for the code generation optimizations in Julia should take a bow!  Please review again!

**Comment by carnaval** [comment]
@vtjnash Don't know why I put `@inline`. Didn't even try it without. It's almost certain it's not necessary. I think it's because `@inline function` is the closest literal translation to `#define` :-)

File: src/support/utf8.h [hunk 3]
```diff
@@ -8,6 +8,15 @@ extern "C" {
...
 #define isutf(c) (((c)&0xC0)!=0x80)
 
+//! Return if a 16-bit or 32-bit character is a surrogate character
+#define	is_surrogate_char(ch)  (((ch) & ~0x7ff) == 0xd800)
+//! Return if a 16-bit or 32-bit character is a leading surrogate character
+#define	is_surrogate_lead(ch)  (((ch) & ~0x3ff) == 0xd800)
+//! Return if a 16-bit or 32-bit character is a trailing surrogate character
+#define	is_surrogate_trail(ch) (((ch) & ~0x3ff) == 0xdc00)
+//! Return if a byte is a valid UTF-8 continuation character
+#define	is_valid_continuation(ch) ((ch & 0xc0) == 0x80)
```
Comment (tkelman) [inline src/support/utf8.h]: Are these `#define`s still necessary? Bit unlikely but these could cause collisions with any other projects that want to include this header.

**Comment by nalimilan** [comment]
> I wonder - could the compiler be smart enough to see that I am using the bounds of the array for the loop, and simply act as if I used @inbounds automatically? Now that would be a really useful optimization!

@ScottPJones My understanding is that in simple cases like iterating over an array the compiler is able to hoist the bounds check out of the loop. You can easily check that by looking at the generating code or by measuring whether `@inbounds` make a significant performance difference.

**Comment by carnaval** [comment]
@nalimilan I think that's a myth. At least I've never seen it happening. Even in the simplest case, the best we(/llvm) do is hoist the load of the array length from the loop. Maybe it works on some LLVM version, or maybe we just need to emit the code a bit differently.

**Comment by ScottPJones** [comment]
>@inline should only be used when strictly necessary. In that case, the function is so simple that inlining should happen automatically anyway.

@nalimilan About the @inline, I just put that into the code because I was using @carnaval 's changes that made the Julia code as fast (usually a good deal faster ;-) ) than the C version.
If only the @inbounds is really necessary, I'll change it to just do that.
>Same here. Also, the spacing of ((byte&0xc0)==0x80) is inconsistent with above, as well as the name of the function.

I didn't write that function... which is why the neither the names nor the spacing are not the same.  Do I really need to go back and clean up somebody else's code?  My names came from my own C code, which predated my involvement with Julia... I had functions in C that worked for 16-bit or 32-bit, and so I did't want "utf16_" or "u16_" as part of the name...

>@ScottPJones My understanding is that in simple cases like iterating over an array the compiler is able to hoist the bounds check out of the loop. You can easily check that by looking at the generating code or by measuring whether @inbounds make a significant performance difference.

That's *exactly* what I already did, and posted the results on the Gist I referenced earlier...

@tkelman I'll go back and check if I can remove those defines... I've been editing a few different branches so fast and furiously, trying to incorporate people's suggestions, that I missed that...  Thanks!

**Comment by vtjnash** [comment]
> I think that's a myth. At least I've never seen it happening. Even in the simplest case, the best we(/llvm) do is hoist the load of the array length from the loop. Maybe it works on some LLVM version, or maybe we just need to emit the code a bit differently.

the checks end up having a low runtime cost on a modern processor architecture, so it may seem like they got removed even while they are still being executed. although, if the presence of the branch inhibits vectorization, then the performance difference can be very noticeable. there is an (experimental?) inductive reasoning pass in llvm for helping to elide these checks automatically, but i don't think we have it enabled right now.

**Comment by carnaval** [comment]
@vtjnash Yes the never taken branch itself is not the problem on most CPU. I'm hopeful there is a permutation of opt. passes that figures it out since, at least in the simple `for x in A` case, the backend understands that the length is constant and that it is the same that is being compared for the bounds check and the loop exit condition. Also the behavior of the induction variable should be completely known by the SCEV infrastructure.

Maybe the problem is the way we formulate the bound check with the single unsigned comparison ?

**Comment by nalimilan** [comment]
@carnaval Thanks, this is a bit disappointing. I thought I had read this several times, but I may have over-interpreted it. 

> I didn't write that function... which is why the neither the names nor the spacing are not the same. Do I really need to go back and clean up somebody else's code? My names came from my own C code, which predated my involvement with Julia... I had functions in C that worked for 16-bit or 32-bit, and so I did't want "utf16_" or "u16_" as part of the name...

@ScottPJones Sorry, I see your code is consistent within utf8.jl. It's just that utf16.jl and utf8.jl use different conventions (which wasn't obvious when I looked at the diff).

**Comment by carnaval** [comment]
I see two problems here. First, we don't hoist the array data pointer load out of the loop. I'll try to understand why. Second, looking at what should be equivalent C to what we'd like to generate for `s=0;for x in A;s+=x;end;s` :

```C
extern void throw_bounds(int*,int*)  __attribute__((noreturn));
int f(int* p, int* plen)
{
    int s = 0;
    int n = *plen;
    int i = 0;
    while (i < n) {
        if (!(i < *plen)) {
            int* x = alloca(sizeof(int));
            *x = i;
            throw_bounds(p,x);
            __builtin_unreachable();
        }
        s += p[i];
        i++;
    }
    return s;
}
```

I can see that, under O3, clang+llvm3.6 suprisingly does _not_ succeed in removing the check, even though the `*plen` load is correctly hoisted by LICM. It's likely to be a pass ordering problem since running `opt -O3` on the resulting bytecode works and vectorizes the loop correctly.

**Comment by tkelman** [comment]
@carnaval if you mean needing to manually hoist the `s.data` getfield, I think that's https://github.com/JuliaLang/julia/issues/9755? Though we should be doing tbaa on immutable types I thought.

**Comment by carnaval** [comment]
The problem here is not an aliasing one I think. The only aliasing imprecision there could be would be on the small dynamic alloca which I believe any simple AA would say is noalias with pretty much anything else.
I'm pretty sure LICM does not hoist the pointer load due to concerns over throw_error() that could be modifying it. I'm not sure though since it also happens when I mark it readonly ... also it should not care since it is an exit block and could never be seen by the rest of the loop ...

**Comment by carnaval** [comment]
Okay so the following :
```C
extern void throw_bounds(int**,int*)  __attribute__((noreturn));
int f(int** A, int* plen)
{
    int s = 0;
    int n = *plen;
    int i = 0;
    while (i < n) {
        int* p = *A;
        if (!(i < *plen)) {
            int* x = alloca(sizeof(int));
            *x = i;
            throw_bounds(A,x);
            __builtin_unreachable();
        }
        s += p[i];
        i++;
    }
    return s;
}
```
taken out of `clang -O0` is optimized fully after two runs under `opt -O3`, one to remove the bound check and the second which then understand it should hoist the pointer load.

Now we just have to figure out which passes are actually useful. For more complicated loop it's a matter of giving precise enough tbaa information that llvm is not confused by ambient stores.

**Comment by JeffBezanson** [comment]
For us, 1d arrays are growable so we don't always hoist loads of the length and data pointer.

**Comment by ScottPJones** [comment]
I have retested with the @inline macros removed, it passed all of the unit tests, as well as being the same speed with my performance test (going from 4 characters up to 4M, with strings with only ASCII, only Latin1, 2-byte UTF-8, 3 byte UTF-8, 4 byte UTF-8, and 6 byte encoded surrogate pairs (which you get from some legacy UTF-8 encoders where UTF-16 wasn't really supported correctly).
Are there any other changes I need to make before this can be merged?
Thanks!

**Comment by quinnj** [comment]
@ScottPJones, please put backticks ( \` \` ) around code, especially macro names (in your last comment: `@inline`). Without backticks, this is actually referencing/notifying/emailing the GitHub user "inline" that he/she was mentioned in this thread.

**Comment by ScottPJones** [comment]
@quinnj  Sorry, someone had already warned me about that, and I've *tried* to be careful... Julia makes it a bit hard though! ;-)

File: src/support/utf8.h [hunk 4]
```diff
@@ -22,8 +22,15 @@ size_t u8_wc_toutf8(char *dest, uint32_t ch);
...
-DLLEXPORT size_t u8_charnum(const char *s, size_t offset);
+/**
+ * @brief      Calculates number of characters in UTF-8 encoded string
+ *
+ * @param[in]  iStr UTF-8 encoded string (or substring)
+ * @param[in]  iLen Length of string (or substring) in bytes
+ *
+ * @return     Number of logical characters (or codepoints) in string (or substring)
+ */
+DLLEXPORT size_t u8_charnum(const char *iStr, size_t iLen);
```
Comment (nalimilan) [inline src/support/utf8.h]: You're using different argument names here. Also, I don't think the docs should be repeated in the headers.

**Comment by nalimilan** [comment]
Thanks, this looks like a nice improvement. Though I'm not sure anything has been decided about the documentation format for C code.

**Comment by carnaval** [comment]
@JeffBezanson I know, but the C code I've written reloads it every time also. In that case it should be quite easy for it to see that the load is loop invariant. I was trying to understand why can't LLVM (really, our set of passes) figure it out on its own.

File: src/support/utf8.h (hunk 4 above)
Comment (ScottPJones) [inline src/support/utf8.h]: OK... I'll clean that up, thanks!

**Comment by ScottPJones** [comment]
>Thanks, this looks like a nice improvement. Though I'm not sure anything has been decided about the documentation format for C code.

@nalimilan About the documentation format... is there a problem with doing *something*? (doxygen with Markdown markup)?  Julia has been around for many years now, and there's been no decision on documenation standards?  Sorry if it seems snarky, it's just very frustrating to deal with code that isn't documented, and from previous experience with doxygen, it does a pretty good job...

**Comment by nalimilan** [comment]
@ScottPJones That's not for me to decide. I agree having a standardized documentation format is a good idea, but better think about it beforehand. :-)

**Comment by tkelman** [comment]
> is there a problem with doing something? (doxygen with Markdown markup)? Julia has been around for many years now, and there's been no decision on documenation standards? Sorry if it seems snarky, it's just very frustrating to deal with code that isn't documented

No there has not been a decision. There have however been consistent, well-warranted complaints that the C code and some core parts of the standard library are under-documented and difficult to understand by anyone who didn't write them. I would actually be highly in favor of a concerted effort to go through and document what's there, if there are people who are willing to help with that. The only issue, and I think it's a minor one, is that there are parts of the codebase that change rapidly, and keeping the documentation up-to-date with the code means a bit of extra work. I do think Julia's a large enough project now with enough people regularly trying to look at and work with the internals that it's past time for that to be a good enough reason to resist documenting things.

**Comment by IainNZ** [comment]
(should move this documentation conversation to a new issue)

**Comment by ScottPJones** [comment]
Would the PTB want me to remove the doxygen style documentation that slipped in?
(that happened because I had originally had that function, that I had written to replace Jeff's, as part of a totally separate C module, that had my C version of UTF conversion functions as well... but since now I have pure Julia versions of the UTF conversion functions, as well as these length functions, the C module is no longer needed)
I'd also be happy to add comments to the files I've been working on (char.jl, utf*.jl), but that's a harder issue, which I already brought up on julia-dev... the `@doc` macro doesn't work for code in Base... it has dependencies on Markdown, which has dependencies all over the place).  For now, in those files, I just put `#=` and `=#` around the whole `@doc` section...
Personally, I feel bad putting in code without documentation... it feels half done...

**Comment by ScottPJones** [comment]
@nalimilan Another thing... doxygen style (using Markdown) comments have already been accepted in the utf8proc module (I hadn't noticed it at first, since it is only in utf8proc.h, and not in utf8proc.c), and that was discussed and accepted... (see issues 26 and 29 in JuliaLang/utf8proc) [btw, how do you directly reference issues that are not in JuliaLang/julia? thx!]

**Comment by pao** [comment]
@ScottPJones `JuliaLang/utf8proc#26`

**Comment by jakebolewski** [comment]
@ScottPJones it would be great if you could organize your performance tests a bit and start contributing some of that code to `test/perf`.  That way we can objectively look at the performance gains over time and make sure there are no regressions as you work on the string code.

**Comment by ScottPJones** [comment]
@jakebolewski Thanks! I wasn't aware of `test/perf`, will investigate now.

**Pull request merged**